File: esexpr/java/src/main/java/dev/argon/esexpr/ESExpr.java
package dev.argon.esexpr;

import org.jetbrains.annotations.NotNull;

import java.math.BigInteger;
import java.util.List;
import java.util.Map;

public sealed interface ESExpr {
    public static record Constructed(@NotNull String constructor, @NotNull Map<String, ? extends ESExpr> kwargs, @NotNull List<? extends @NotNull ESExpr> arguments) implements ESExpr {}

    public static record Bool(boolean b) implements ESExpr {}

    public static record Int(@NotNull BigInteger n) implements ESExpr {}

    public static record Str(@NotNull String s) implements ESExpr {}

    public static record Binary(byte @NotNull[] b) implements ESExpr {}

    public static record Float32(float f) implements ESExpr {}

    public static record Float64(double d) implements ESExpr {}


    public static record Null() implements ESExpr {}

}

File: esexpr/src/main/scala/dev/argon/esexpr/ESExpr.scala
package dev.argon.esexpr

import zio.Chunk

enum ESExpr derives CanEqual {
  case Constructed(constructor: String, kwargs: Map[String, ESExpr], args: Seq[ESExpr])
  case Bool(b: Boolean)
  case Int(n: BigInt)
  case Str(s: String)
  case Binary(b: Chunk[Byte])
  case Float32(f: Float)
  case Float64(d: Double)
  case Null
}
